File: Python3/22.Generate_Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        self.helper("", n, 0, ans)
        return ans
        
    def helper(self, prefix, num_starting, num_ending, ans):
        if (num_starting == 0 and num_ending == 0):
            ans.append(prefix)
            return
        
        if (num_starting != 0):
            self.helper(prefix + '(', num_starting - 1, num_ending + 1, ans)
        if (num_ending != 0):
            self.helper(prefix + ')', num_starting, num_ending - 1, ans)

File: Java/22.Generate_Parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        helper(n, 0, new StringBuilder(), ans);
        return ans;
        
    }
    
    private void helper(int open, int close, StringBuilder prefix, List<String> ans) {
        if (open == 0 && close == 0) {
            ans.add(prefix.toString());
            return;
        }

        if (open != 0) {
            helper(open - 1, close + 1, prefix.append('('), ans);
            prefix.deleteCharAt(prefix.length() - 1);
        }
        if (close != 0) {
            helper(open, close - 1, prefix.append(')'), ans);
            prefix.deleteCharAt(prefix.length() - 1);
        }
    }
}